creator/ingest_runs/genomic_data_loader.py

```diff
@@ -0,0 +1,356 @@
+from creator.ingest_runs import utils
+import ast
+import os
+import logging
+from pprint import pprint
+
+import pandas as pd
+
+from kf_lib_data_ingest.app import settings as ingest_settings
+from kf_lib_data_ingest.common.concept_schema import CONCEPT
+from kf_lib_data_ingest.etl.load.load import LoadStage
+
+from creator.ingest_runs.data_generator.ingest_package.extract_configs.s3_scrape_config import genomic_file_ext  # noqa
+
+GEN_FILE = "genomic_file"
+GEN_FILES = "genomic-files"
+SEQ_EXP = "sequencing_experiment"
+SEQ_EXPS = "sequencing-experiments"
+SEQ_EXP_GEN_FILE = "sequencing_experiment_genomic_file"
+SEQ_EXP_GEN_FILES = "sequencing-experiment-genomic-files"
+BIO_GEN_FILE = "biospecimen_genomic_file"
+BIO_GEN_FILES = "biospecimen-genomic-files"
+LOAD_ENTITY_TYPES = [GEN_FILE, BIO_GEN_FILE, SEQ_EXP_GEN_FILE]
+
+LOCALHOST = "http://localhost:5000"  # testing
+DATASERVICE_URL = LOCALHOST  # During testing
+
+
+class GenomicDataLoader(object):
+
+    def __init__(self, study_id, init_logger=False):
+        self.study_id = study_id
+
+        if init_logger:
+            format_ = (
+                "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+            )
+            root = logging.getLogger()
+            consoleHandler = logging.StreamHandler()
+            consoleHandler.setFormatter(logging.Formatter(format_))
+            root.setLevel("INFO")
+            root.addHandler(consoleHandler)
+
+        self.logger = logging.getLogger(type(self).__name__)
+
+        self.loader = None
+        self.target_api_cfg = ingest_settings.load().TARGET_API_CONFIG
+
+    def ingest_gwo(self, manifest_df):
```

Comment:
Looking through this function, I can see why testing is going to be headache.
Consider all the dependencies needed to test this single function. From what I can see, you need:
1) A dataframe with the right data (It's difficult for me to figure out what that is)
2) Some sort of s3 buckets need to be setup in the correct manner
3) Genomic files need to be put in the s3 buckets
4) A manifest that references the genomic files needs to be placed in the s3 buckets
5) A dataservice that is populated with the required genomic files
6) A dataservice that is populated with the required sequencing experiments
7) A dataservice that relates genomic files to sequencing experiments as needed

When I run into a piece of code that requires such a vast amount of dependencies, it usually tells me that we can find better abstractions. If we can formulate such abstractions such that we only need to have one or two dependencies in place, it's going to make our life much easier when it comes to testing. Another benefit may be that we get some handy abstractions that we may use in other places.

If I may hint at what sort of abstractions may help here, let me first suggest considering how this may be approached _without_ the use of DataFrames and pandas. DataFrames are themselves useful abstractions because they allow quick analysis of data by establishing very general concepts about what you can do with data. This allows someone in an interactive environment to do common data operations such as reshape, filter, and add fields on a huge space of different data values. This is extremely powerful when the need is to explore and analyze data quickly, but this isn't quite the same need that we have. In fact, the need we have here is quite the opposite.

Here, we are trying to take a pretty rigid list of data (the GWO) and map it into another rigid datamodel (the genomic files and such in the Dataservice). It may seem handy that DataFrames offer batch operations to reshape data, but I would suggest that modeling and operating on our data on a per-instance basis is much closer to how it exists in reality (the dataservice is implemented in this fashion after all).

Consider this design:
If instead of a generic dataframe, you had a list of `GenomicFile`s. Each `GenomicFile` would have your standard properties that you see in Dataservice. There are then properties that depend on other properties such as how `size` needs `url` to resolve what the size of the file is. This 'size' field would probably have some implementation that looks at the protocol of the `url` and uses the appropriate resolver to find the value to return.

Now let's consider what we'd have to do to test this design:
First, we would probably want to  test that our `size` field returns the correct value. We would write `test_genomic_file_size()` where we would make sure this gets done right for each protocol. If we wanted to further this abstraction, we could write `FileResolver`s or similar that get invoked based on the protocol and we would then want to test each `FileResolver`. Writing these tests would require that we model each dependency and simulate the different possible results. For the `HTTPFile`, say, we just need to check what happens when the file gets resolved and what happens when a 404 is returned. The beauty here is that we only have this one straightforward dependency to mock.
As far as other tests and mocking, it may at first seem like we will need to mock out these resolvers whenever we want to access the `size` property so we don't try to make network requests during the test. However, because of the work we did to formulate the `FileResolver` abstraction,  it is almost trivial to implement something like a `SelfResolver` which stores the size on the `GenomicFile` and thus no network requests would be needed. We would then be able to use this resolver in all our tests and then have zero required dependencies when we test other features utilizing `GenomicFile`s.

One last point: I see that there does appear to be a requirement to provide a `DataFrame` to some of the ingest library functions. Perhaps this may seem to promote using `DataFrame`s more, but recall that we can easily adjust our abstraction to handle this. For example, perhaps we add a `to_df()` class method or similar to make that conversion just in time before the hand-off to ingest lib.

So finally, I hope this shows that re-framing to a more appropriate conceptual model can save tons of headache when testing. It should also go to show that it can allow new functionality to be added very rapidly (ie adding a new `FileResolver`). I think we could spend a bit of time to re-think the design here to set ourselves up for future work and to save ourselves from crazy testing hell.